oak/server/rust/oak_runtime/src/message.rs

```diff
@@ -14,12 +14,21 @@
 // limitations under the License.
 //
 
-use crate::Handle;
-
 /// Encapsulates a message consisting of opaque data bytes and a vector of channels.
-/// The data bytes should not contain any pointers or handles.
+/// The data bytes should not contain any pointers or handles.  Note that `Message`
+/// and `Channel` objects can be leaked if the Oak application creates cycles of
+/// references (e.g. the only reference to a `Channel` is in a `Message` that is
+/// held in the same `Channel`).
 #[derive(Debug)]
 pub struct Message {
     pub data: Vec<u8>,
-    pub channels: Vec<Handle>,
+    pub channels: Vec<crate::runtime::ChannelHalf>,
+}
+
+/// Encapsulates a message relative to a specific Node, consisting of opaque data bytes
+/// together with a vector of ABI handles that are relative to the Node.
+#[derive(Debug)]
+pub struct NodeMessage {
+    pub data: Vec<u8>,
+    pub handles: Vec<oak_abi::Handle>,
```

Comment:
I understand the need for both these types, because we remap handles within the handle space of each node; I wonder though if we should consider a simpler approach in which we just use one global numbering space, and we don't renumber handles when we pass them around. We still have the runtime track bytes and handles separately, and only allow using them if they have been granted to the node that is trying to use them.

Not for this PR in any case, but what's your view on that @daviddrysdale ?

edit: I guess actually a good reason for **not** doing that is that if we added first class support for rights associated with each handle, then we need the extra layer of indirection so that we may have two handles pointing to the same underlying object, but with different rights, so that it may be passed to different nodes